openedx/core/lib/xblock_runtime/blockstore_kvs.py

```diff
@@ -0,0 +1,148 @@
+
+# In order to work around this situation without changing the API, we
+# require all field reads and/or writes to be part of a
+# blockstore_transaction() context; that context is mostly used to know
+# when we're using field data, and when it exits the in-memory cache
+# is erased. More importantly, all field _writes_ must be wrapped in
+# either a collect_changes() or a collect_parsed_fields() context.
+
+# The collect_parsed_fields() context should be used when parsing,
+# and it means that writes will be considered to be reads in disguise
+# (we're setting the value in memory to match the value in the XML).
+
+# The collect_changes() context should be used when running any
+# XBlock view or handler, and it signifies that writes are deliberate
+# changes.
+
+_cached_keys = {}  # Values that we read out of XML during a transaction, and/or committed writes
+_transaction_depth = 0  # This is > 0 inside a blockstore_transaction() context
+_pending_writes = []  # Values that have been written by the XBlock but not yet persisted
+
+
+@contextmanager
+def blockstore_transaction():
+    global _transaction_depth
+    _transaction_depth = _transaction_depth + 1
+    try:
+        yield
+    except:
+        raise
+    finally:
+        _transaction_depth = _transaction_depth - 1
+        if _transaction_depth == 0:
+            _cached_keys.clear()
+
+
+@contextmanager
+def collect_parsed_fields():
+    """
+    Use this context manager while parsing XBlock XML.
+    All field writes are assumed to be converting data
+    from XML to Python, and NOT reflecting deliberate
+    changes made to field values.
+    """
+    if _transaction_depth < 1:
+        raise RuntimeError("You need to be in a blockstore_transaction() context.")
+    new_keys = {}
+    _pending_writes.append(new_keys)
+    try:
+        yield
+        # Now save all new_keys into _cached_keys:
+        for key, value in new_keys.iteritems():
+            _cached_keys[key] = value
+    except:
+        raise
+    finally:
+        _pending_writes.pop()  # Discard all changes
+
+@contextmanager
+def collect_changes():
+    """
+    Use this context manager while running an XBlock view or handler that
+    may be saving new field values. Changes will be collected and then
+    when this context manager exists, the changes will be persisted to
+    blockstore in an atomic transaction.
+    """
+    if _transaction_depth < 1:
+        raise RuntimeError("You need to be in a blockstore_transaction() context.")
+    new_keys = {}
+    _pending_writes.append(new_keys)
+    try:
+        yield
+        # TODO: Now persist all changes to Blockstore
+        # Now save all new_keys into _cached_keys:
+        for key, value in new_keys.iteritems():
+            _cached_keys[key] = value
+    except:
+        raise
+    finally:
+        _pending_writes.pop()  # Discard all changes
+
+
+class BlockstoreKVS(InheritanceKeyValueStore):
```

Comment:
Are we supporting inheritance then?

Reply:
I'm not sure yet; I could just put the normal keyvaluestore base class here for now if you think that's better.

Actually, yeah, that's probably better; because indeed I wasn't intending to support inheritance, but rather that concept I've mentioned in the past where the course can define a list of rules like "all `problem` blocks in this course should default to `allow_retries=3`" but there's a single flat list of rules, not a complex inheritance pathway.